clients/drcachesim/tools/trace_filter.cpp

```diff
@@ -0,0 +1,217 @@
+        return false;
+    }
+    if (!directory_iterator_t::is_directory(trace_dir) ||
+        !directory_iterator_t::is_directory(output_dir)) {
+        ERRMSG("Trace dir or output dir is not a directory\n");
+        return false;
+    }
+    directory_iterator_t end;
+    directory_iterator_t iter(trace_dir);
+    if (!iter) {
+        ERRMSG("Failed to list directory %s: %s", trace_dir.c_str(),
+               iter.error_string().c_str());
+        return false;
+    }
+    for (; iter != end; ++iter) {
+        const std::string fname = *iter;
+        if (fname == "." || fname == "..")
+            continue;
+        const std::string trace_path = trace_dir + DIRSEP + fname;
+        const std::string output_path = output_dir + DIRSEP + fname;
+        std::unique_ptr<trace_entry_reader_t> reader = get_reader(trace_path, verbosity);
+        if (!reader) {
+            error_string_ = "Could not get a reader for " + trace_path;
+            return false;
+        }
+
+        std::unique_ptr<std::ostream> writer = get_writer(output_path, verbosity);
+        if (!writer) {
+            error_string_ = "Could not get a writer for " + output_path;
+            return false;
+        }
+
+        thread_data_.push_back(shard_data_t(static_cast<int>(thread_data_.size()),
+                                            std::move(reader), std::move(writer),
+                                            trace_path, output_path));
+    }
+    // Like raw2trace, we use a simple round-robin static work assigment.  This
+    // could be improved later with dynamic work queue for better load balancing.
+    if (worker_count_ <= 0)
+        worker_count_ = std::thread::hardware_concurrency();
+    worker_tasks_.resize(worker_count_);
+    int worker = 0;
+    for (size_t i = 0; i < thread_data_.size(); ++i) {
+        VPRINT(this, 2, "Worker %d assigned trace shard %zd\n", worker, i);
+        worker_tasks_[worker].push_back(&thread_data_[i]);
+        thread_data_[i].worker = worker;
+        worker = (worker + 1) % worker_count_;
+    }
+    trace_end_ = std::unique_ptr<default_trace_entry_file_reader_t>(
+        new default_trace_entry_file_reader_t());
+    return true;
+}
+
+void
+trace_filter_t::process_tasks(std::vector<shard_data_t *> *tasks)
+{
+    if (tasks->empty()) {
+        VPRINT(this, 1, "Worker has no tasks\n");
+        return;
+    }
+    VPRINT(this, 1, "Worker %d assigned %zd task(s)\n", (*tasks)[0]->worker,
+           tasks->size());
+    for (shard_data_t *tdata : *tasks) {
+        VPRINT(this, 1, "Worker %d starting on trace shard %d\n", tdata->worker,
+               tdata->index);
+        if (!tdata->iter->init()) {
+            tdata->error = "Failed to read from trace" + tdata->trace_file;
+            return;
+        }
+        for (; *tdata->iter != *trace_end_; ++(*tdata->iter)) {
+            const trace_entry_t &entry = **tdata->iter;
+            // TODO i#5675: Filter entries here.
```

Comment:
I think we want a framework like for regular trace processing so we can create separate tools rather than hardcoding different filters here.  Even if we anticipate fewer such tools than regular analysis tools, such a setup conforms to the existing analysis tool setup better and seems like it would actually avoid code duplication and so in that sense be simpler in the long run maintenance-wise.

Reply:
You mean separate tools for filtering? Having a separate sub-class for each kind of filter would probably add boilerplate. I was thinking of adding each filter as a bool-returning-routine that is toggled using a command-line flag. As you noted and as discussed offline previously, regular trace analysis would continue to be on memref_t and this is better off as a standalone tool.

Reply:
Converted analyzer_t to a template in https://github.com/DynamoRIO/dynamorio/pull/5720.